### PR TITLE
JSON-API: Allow `close` channel to specified address

### LINKED ISSUE #2027: Close channel to specified address
Add new parameter to "lightning-cli close  {destination} address {satoshi} (or 'all')" to have opportunitie close LN channel to any on-chain wallet (use LN channel funds for on-chain payments). Remaining funds should be sent to LN wallet.

### PR DESCRIPTION
**Pull request by trueptolemy**
Fixed #2027
Though we haven't supported local `upfront_shutdown_script`, we can add an optional patameter `destination` in `close`, to which the to-local output will be sent. (In the future, when we set the local `upfront_shutdown_script`, if `destination` is different with the local `upfront_shutdown_script`, `close` will fail.)

About the `satoshi` parameter in #2027, I think it isn't allowed in spec.

### CODE REVIEW

**Comment by trueptolemy** [comment]
Related with rusty's [comment](https://github.com/ElementsProject/lightning/pull/3073#discussion_r329379760), here's the same issue...continue to fix...

**Comment by trueptolemy** [comment]
I meet this error:

==15220== Invalid write of size 8
==15220==    at 0x111572: bitcoin_tx_output_set_amount (tx.c:184)
==15220==    by 0x110E5C: bitcoin_tx_add_output (tx.c:50)
==15220==    by 0x130163: funding_tx (funding_tx.c:42)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220==  Address 0x564cfd0 is 16 bytes before a block of size 32 alloc'd
==15220==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15220==    by 0x12B532: allocate (tal.c:245)
==15220==    by 0x12B68B: add_notifier_property (tal.c:300)
==15220==    by 0x12BE8B: tal_add_destructor_ (tal.c:516)
==15220==    by 0x111FDA: bitcoin_tx (tx.c:392)
==15220==    by 0x131B72: tx_spending_utxos (utxo.c:72)
==15220==    by 0x13009D: funding_tx (funding_tx.c:29)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220== 
==15220== Invalid read of size 8
==15220==    at 0x131519: output_better (permute_tx.c:118)
==15220==    by 0x131695: find_best_out (permute_tx.c:152)
==15220==    by 0x131764: permute_outputs (permute_tx.c:172)
==15220==    by 0x130178: funding_tx (funding_tx.c:44)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220==  Address 0x564cfd0 is 16 bytes before a block of size 32 alloc'd
==15220==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15220==    by 0x12B532: allocate (tal.c:245)
==15220==    by 0x12B68B: add_notifier_property (tal.c:300)
==15220==    by 0x12BE8B: tal_add_destructor_ (tal.c:516)
==15220==    by 0x111FDA: bitcoin_tx (tx.c:392)
==15220==    by 0x131B72: tx_spending_utxos (utxo.c:72)
==15220==    by 0x13009D: funding_tx (funding_tx.c:29)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220== 
==15220== Invalid read of size 8
==15220==    at 0x13152C: output_better (permute_tx.c:119)
==15220==    by 0x131695: find_best_out (permute_tx.c:152)
==15220==    by 0x131764: permute_outputs (permute_tx.c:172)
==15220==    by 0x130178: funding_tx (funding_tx.c:44)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220==  Address 0x564cfd0 is 16 bytes before a block of size 32 alloc'd
==15220==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15220==    by 0x12B532: allocate (tal.c:245)
==15220==    by 0x12B68B: add_notifier_property (tal.c:300)
==15220==    by 0x12BE8B: tal_add_destructor_ (tal.c:516)
==15220==    by 0x111FDA: bitcoin_tx (tx.c:392)
==15220==    by 0x131B72: tx_spending_utxos (utxo.c:72)
==15220==    by 0x13009D: funding_tx (funding_tx.c:29)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220== 
==15220== Invalid read of size 8
==15220==    at 0x1117BD: bitcoin_tx_output_get_amount (tx.c:229)
==15220==    by 0x111065: bitcoin_tx_compute_fee (tx.c:88)
==15220==    by 0x11116F: elements_tx_add_fee_output (tx.c:102)
==15220==    by 0x1301C0: funding_tx (funding_tx.c:52)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220==  Address 0x564cfd0 is 16 bytes before a block of size 32 alloc'd
==15220==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15220==    by 0x12B532: allocate (tal.c:245)
==15220==    by 0x12B68B: add_notifier_property (tal.c:300)
==15220==    by 0x12BE8B: tal_add_destructor_ (tal.c:516)
==15220==    by 0x111FDA: bitcoin_tx (tx.c:392)
==15220==    by 0x131B72: tx_spending_utxos (utxo.c:72)
==15220==    by 0x13009D: funding_tx (funding_tx.c:29)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220== 
==15220== Invalid read of size 8
==15220==    at 0x1117BD: bitcoin_tx_output_get_amount (tx.c:229)
==15220==    by 0x1323A5: main (run-funding_tx.c:188)
==15220==  Address 0x564cfd0 is 16 bytes before a block of size 32 alloc'd
==15220==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15220==    by 0x12B532: allocate (tal.c:245)
==15220==    by 0x12B68B: add_notifier_property (tal.c:300)
==15220==    by 0x12BE8B: tal_add_destructor_ (tal.c:516)
==15220==    by 0x111FDA: bitcoin_tx (tx.c:392)
==15220==    by 0x131B72: tx_spending_utxos (utxo.c:72)
==15220==    by 0x13009D: funding_tx (funding_tx.c:29)
==15220==    by 0x13233D: main (run-funding_tx.c:178)
==15220== 
Makefile:443: recipe for target 'unittest/common/test/run-funding_tx' failed
make: *** [unittest/common/test/run-funding_tx] Error 7

**Comment by trueptolemy** [comment]
> Pull request builds check the following cache locations in order, using the first one present:
> 
> The pull request cache.
> The pull request target branch cache.
> The repository default branch cache.

Well, after some tries, I think this is because of the cache of this PR... @cdecker could you please delete the cache? :-) 
Then if passed, I would close #3128 .

**Comment by justinmoon** [comment]
Proviso: I'm not a C programmer

This would be an extremely helpful feature for efforts to make a [plugin](https://github.com/justinmoon/clhwi) to utilize HWI to open and close channels using hardware wallets.

I successfully built this PR, then spent an hour screwing around with [lnet](https://github.com/cdecker/lnet) trying to get a setup to test this feature on regtest. Somewhere along the line I broke it:

```
$ ./lightningd/lightningd 
lightningd: FATAL SIGNAL 11 (version v0.7.2.1-424-gd72bfad)
0x5572776daaf8 send_backtrace
	common/daemon.c:40
0x5572776dab83 crashdump
	common/daemon.c:53
0x7fdf75d547df ???
	???:0
0x7fdf75e7bc67 ???
	???:0
0x5572776ecfa6 towire
	wire/towire.c:21
0x5572776ed45a towire_sha256
	wire/towire.c:172
0x5572776ed468 towire_sha256_double
	wire/towire.c:177
0x5572776ed4dc towire_bitcoin_blkid
	wire/towire.c:194
0x5572776ed661 towire_chainparams
	wire/towire.c:262
0x5572776e4165 towire_hsm_init
	hsmd/gen_hsm_wire.c:103
0x5572776b91af hsm_init
	lightningd/hsm_control.c:96
0x5572776bfc72 main
	lightningd/lightningd.c:712
0x7fdf75d40ee2 ???
	???:0
0x5572776af29d ???
	???:0
0xffffffffffffffff ???
	???:0
Segmentation fault (core dumped)
```

Attempting to rebuild with `make clean` followed by `make` doesn't help. Let me know how to resolve this. Again, I'm not familiar with C or c-lightning :grimacing: so I apologize for the noobery in advance!

**Comment by rsbondi** [comment]
I don't know if this is relevant, but here is what I discovered, this works for me if I say

`l1-cli close 02fabc... 0 bcrt1qf...`

but I have always been able to ignore optional parameters with `null` so if I try

`l1-cli close 02fabc... null bcrt1qf...`

I expect the null to ignore the `unilateraltimeout` but the `destination` is also ignored

I don't know if this is an actual convention, or if I just have gotten lucky up until now, but I have used `null` in this way many times with success.  Doing so here, I get the txid returned, but the `destination` address used does not show in the transaction and the funds do get returned to the internal wallet.

Maybe it is just a matter of me assuming a convention that does not really exist but just happens to work in other cases.

**Comment by darosior** [comment]
> Maybe it is just a matter of me assuming a convention that does not really exist but just happens to work in other cases.

I think so :-). Just use the `-k` CLI parameter to pass named arguments and thus be able to pass the argument `n` without passing argument `n-1`.

**Comment by niftynei** [comment]
fyi travis CI is complaining that you have a valgrind error in `run-bolt11`

```
VALGRIND=1 valgrind -q --error-exitcode=7 --track-origins=yes --leak-check=full --show-reachable=yes --errors-for-leak-kinds=all common/test/run-bolt11 > /dev/null
```

**Comment by trueptolemy** [comment]
```
but I have always been able to ignore optional parameters with null so if I try

l1-cli close 02fabc... null bcrt1qf...

I expect the null to ignore the unilateraltimeout but the destination is also ignored
```

@rsbondi Thank you! :-) 
I haven't thought about this form, and I also use the `CLI` like what @darosior  said with the `-k`! I'm not familiar with the `CLI`, maybe @rustyrussell @niftynei  can confirm if your form is valid? Or it can be a new feature for our `CLI`?


And in my understanding, if we want this work:
> `l1-cli close 02fabc... null bcrt1qf... `

`null` will need to be transformed to `"null"`, and clightning need to be able to prase the `"null"`.


For this PR, my idea is:
If `destination` isn't specified, by default the satoshi will be sent to the internal wallet of `clightning`.
Only when you specify the `destination` field, the satoshi will be sent to your specified address.

**Comment by trueptolemy** [comment]
@niftynei  I think the change in this PR doesn‘t’ influence `run-bolt11` or `run-funding_tx`. 
Could you please help me to check if this PR has the cache on travis-CI and delete it? I guess this error is related with the bug of `wally-core` appeared a few days ago, and the cache isn't completely clean up...

**Comment by justinmoon** [comment]
I was able to successfully close channels to legacy, p2sh-segwit, and native segwit addresses on a local regtest network. Haven't tried testnet or mainnet.

I agree with @rsbondi that the `null` value behavior for `unilateraltimeout` is unexpected (I'm a c-lightning noob). Other than that it worked great in my limited testing.

File: lightningd/peer_control.c [hunk 1]
```diff
@@ -1230,44 +1287,75 @@ static struct command_result *json_close(struct command *cmd,
...
 	struct peer *peer;
 	struct channel *channel COMPILER_WANTS_INIT("gcc 7.3.0 fails, 8.3 OK");
-	unsigned int *timeout;
+	unsigned int *timeout = NULL;
 	bool force = true;
 	bool do_timeout;
+	const u8 *local_shutdown_script = NULL;
+	unsigned int *old_timeout;
+	bool *old_force;
+	bool old_style = false;
```
Comment (niftynei) [inline lightningd/peer_control.c]: `old_style` should not be initialized here; initialize it where we know its value.

File: lightningd/peer_control.c [hunk 2]
```diff
@@ -1230,44 +1287,75 @@ static struct command_result *json_close(struct command *cmd,
...
-				/* Old default timeout */
-				if (!timeout) {
-					timeout = tal(cmd, unsigned int);
-					*timeout = 30;
-				}
+			if (json_to_bool(buffer, firsttok, &force))
+				old_style = true;
 			/* New-style timeout */
-			} else {
+			else {
```
Comment (niftynei) [inline lightningd/peer_control.c]: `old_style` is false here.

File: lightningd/peer_control.c [hunk 3]
```diff
@@ -1230,44 +1287,75 @@ static struct command_result *json_close(struct command *cmd,
...
+
+			if (old_style) {
+				timeout = tal(cmd, unsigned int);
+				/* Old default timeout */
+				if (!secondtok)
+					*timeout = 30;
+				else {
+					if (!json_to_number(buffer, secondtok, timeout))
+						return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+								    "(old-style)close: "
```
Comment (niftynei) [inline lightningd/peer_control.c]: `old-style` means nothing to the user of this CLI command. we should remove it from the user facing messages.

(old-style)close: -> close:

File: lightningd/peer_control.c [hunk 4]
```diff
@@ -1230,44 +1287,75 @@ static struct command_result *json_close(struct command *cmd,
...
+								    "Expected timeout to be a number. "
+								    "This old-style is deprecated!");
```
Comment (niftynei) [inline lightningd/peer_control.c]: old-style -> argument ordering

File: lightningd/peer_control.c [hunk 5]
```diff
@@ -1230,44 +1287,75 @@ static struct command_result *json_close(struct command *cmd,
...
-			   p_opt("timeout", param_number, &timeout),
+			   p_opt("unilateraltimeout_or_force",
+				 param_tok_timeout_or_force, &firsttok),
+			   p_opt("destination_or_timeout",
+				 param_tok_dest_or_timeout, &secondtok),
 			   NULL))
 			return command_param_failed();
 
-		if (tok) {
+		if (firsttok) {
```
Comment (niftynei) [inline lightningd/peer_control.c]: seems like you're missing the case where `!firsttok && secondtok` is true

File: lightningd/peer_control.c [hunk 6]
```diff
@@ -1230,44 +1287,75 @@ static struct command_result *json_close(struct command *cmd,
...
+					if (res == ADDRESS_PARSE_UNRECOGNIZED)
+						return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+								    "Could not parse destination address");
+					else if (res == ADDRESS_PARSE_WRONG_NETWORK)
+						return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+								    "Destination address is not on network %s",
+								    get_chainparams(cmd->ld)->network_name);
 				}
 			}
 		}
```
Comment (niftynei) [inline lightningd/peer_control.c]: i'd expect an `else if (secondtok)` here. this should be triggered in the case `close id destination`

File: lightningd/peer_control.c [hunk 7]
```diff
@@ -1336,31 +1431,107 @@ static struct command_result *json_close(struct command *cmd,
...
 		if (channel->owner)
 			subd_send_msg(channel->owner,
-				      take(towire_channel_send_shutdown(channel)));
-	}
+				      take(towire_channel_send_shutdown(NULL,
+					   channel->shutdown_scriptpubkey[LOCAL])));
+	} else if (channel->state == CHANNELD_SHUTTING_DOWN) {
+		/* FIXME: Add to spec that we must allow repeated shutdown! */
+		if (local_shutdown_script)
+			log_debug(channel->log, "Repeated close command: "
```
Comment (niftynei) [inline lightningd/peer_control.c]: only log this if they're different?

File: lightningd/peer_control.c [hunk 8]
```diff
@@ -1336,31 +1431,107 @@ static struct command_result *json_close(struct command *cmd,
...
+				  "the new local scriptpubkey is %s, "
+				  "and the old local scriptpubkey is %s",
+				  local_shutdown_script,
+				  channel->shutdown_scriptpubkey[LOCAL]);
+		else {
+			local_shutdown_script = p2wpkh_for_keyidx(channel,
+								  cmd->ld,
+								  channel->final_key_idx);
+			log_debug(channel->log, "Repeated close command: "
```
Comment (niftynei) [inline lightningd/peer_control.c]: only log this if they're different?

File: lightningd/peer_control.c [hunk 9]
```diff
@@ -1336,31 +1431,107 @@ static struct command_result *json_close(struct command *cmd,
...
+				return command_fail(cmd, LIGHTNINGD,
+						    "Channel is closing now (in state %s) "
+						    "with to-local destination %s",
+						    channel_state_name(channel),
+						    is_default ?
+						    tal_fmt(tmpctx, "(default) %s",
+							    channel->shutdown_scriptpubkey[LOCAL]) :
+						    (char *)channel->shutdown_scriptpubkey[LOCAL]);
+		} else {
+			if (!is_default)
```
Comment (niftynei) [inline lightningd/peer_control.c]: you can merge these two lines

File: tests/test_closing.py [hunk 10]
```diff
@@ -297,6 +297,58 @@ def test_closing_negotiation_reconnect(node_factory, bitcoind):
...
+@unittest.skipIf(not DEVELOPER, "needs DEVELOPER=1")
+def test_closing_specified_destination(node_factory, bitcoind):
+    l1, l2 = node_factory.line_graph(2)
+    chan = l1.get_channel_scid(l2)
+
+    l1.pay(l2, 200000000)
+
+    bitcoind.generate_block(5)
+    addr = 'bcrt1qeyyk6sl5pr49ycpqyckvmttus5ttj25pd0zpvg'
+    l1.rpc.close(chan, destination=addr)
```
Comment (niftynei) [inline tests/test_closing.py]: is there a way to check ordered params in addition to key-value ones with the python rpc? this really needs a test for the single value 'destination' ie

```
close id destination
```

File: doc/lightning-close.7.md [hunk 11]
```diff
@@ -24,6 +25,9 @@ If *unilateraltimeout* is zero, then the **close** command will wait
 indefinitely until the peer is online and can negotiate a mutual close.
 The default is 2 days (172800 seconds).
 
+The *destination* can be of any Bitcoin accepted type, including bech32.
+If it isn't specified, the default is the address of c-lightning.
```
Comment (niftynei) [inline doc/lightning-close.7.md]: the address of c-lightning. -> a c-lightning wallet address

File: CHANGELOG.md [hunk 12]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   and has `connected` and `state` fields for channels, like `listpeers`.
 - JSON API: `fundchannel_start` now includes field `scriptpubkey`
 - JSON API: `txprepare` and `withdraw` now accept an optional parameter `utxos`, a list of utxos to include in the prepared transaction
+- JSON_API: `close` now accept an optional patameter `destination`, to which the to-local output will be sent.
```
Comment (niftynei) [inline CHANGELOG.md]: patameter -> parameter

File: tests/test_closing.py [hunk 13]
```diff
@@ -349,6 +349,37 @@ def test_closing_specified_destination(node_factory, bitcoind):
...
+    l1.rpc.check(command_to_check='close', id=nodeid, destination=addr)
+    # Old-style
+    l1.rpc.check(command_to_check='close', id=nodeid, force=False)
+    l1.rpc.check(command_to_check='close', id=nodeid, force=False, timeout=10)
+    l1.rpc.check(command_to_check='close', id=nodeid, timeout=10)
+
+    l1.rpc.call('check', ['close', nodeid])
+    # Array(new-style)
+    l1.rpc.call('check', ['close', nodeid, 10])
+    l1.rpc.call('check', ['close', nodeid, 0, addr])
```
Comment (niftynei) [inline tests/test_closing.py]: missing check for `l1.rpc.call('check', ['close', nodeid, addr])`

File: lightningd/peer_control.c (hunk 6 above)
Comment (niftynei) [reply lightningd/peer_control.c]: or `close id timeout`?

File: tests/test_closing.py [hunk 14]
```diff
@@ -297,6 +297,89 @@ def test_closing_negotiation_reconnect(node_factory, bitcoind):
...
+    # Array(old-style)
+    l1.rpc.call('check', ['close', nodeid, True, 10])
+    l1.rpc.call('check', ['close', nodeid, False])
```
Comment (niftynei) [inline tests/test_closing.py]: also missing check for `l1.rpc.call('check', ['close', nodeid, 10])`

File: tests/test_closing.py (hunk 10 above)
Comment (rustyrussell) [reply tests/test_closing.py]: Yes, you can feed it directly like so:

```python
l1.rpc.call('close', [chan, addr])
l1.rpc.call('close', {'destination': addr})
```

Comment (trueptolemy) [reply tests/test_closing.py]: I’ll add this test in the tests for compat.

Comment (trueptolemy) [reply tests/test_closing.py]: @niftynei @rustyrussell I find `l1.rpc.call('close', [chan, addr])` doesn't work. It should be `l1.rpc.call('close', [chan, None, addr])` like @rsbondi said.
And I also test the form of `fundchannel_start`:
PASSED:        l1.rpc.call("fundchannel_start", [l2.info['id'], 10**6, None, True])
FAILED:         l1.rpc.call("fundchannel_start", [l3.info['id'], 10**6, True])

The result of `CLI` is the same. (replace `None` to `null`)

**Comment by trueptolemy** [comment]
Thanks you for your review :-) ! @niftynei I fixed the bug and add the tests .

File: channeld/channel_wire.csv [hunk 15]
```diff
@@ -159,6 +159,8 @@ msgtype,channel_got_revoke_reply,1122
 
 # Tell peer to shut down channel.
 msgtype,channel_send_shutdown,1023
+msgdata,channel_send_shutdown,shutdown_len,u16,
```
Comment (darosior) [inline channeld/channel_wire.csv]: This parameter is not reflected in the mocks nor in your usage of `fromwire` and `towire` ?

File: contrib/pylightning/lightning/lightning.py [hunk 16]
```diff
@@ -329,31 +329,29 @@ def _deprecated_close(self, peer_id, force=None, timeout=None):
...
+        the to-local output will be sent to {destination}.
 
         Deprecated usage has {force} and {timeout} args.
         """
-        unilateraltimeout = None
 
         if 'force' in kwargs or 'timeout' in kwargs:
             return self._deprecated_close(peer_id, *args, **kwargs)
 
         # Single arg is ambigious.
```
Comment (darosior) [inline contrib/pylightning/lightning/lightning.py]: This does not check a single arg anymore

File: contrib/pylightning/lightning/lightning.py [hunk 17]
```diff
@@ -329,31 +329,29 @@ def _deprecated_close(self, peer_id, force=None, timeout=None):
...
+                "destination": destination
+            }
+            return self.call("close", payload)
 
-        payload = {
-            "id": peer_id,
-            "unilateraltimeout": unilateraltimeout
-        }
-        return self.call("close", payload)
+        return _close(peer_id, *args, **kwargs)
```
Comment (darosior) [inline contrib/pylightning/lightning/lightning.py]: Why do you use an embedded function ? I think you could use (did not test though)
```python
payload = {
    "id": peer_id,
    "unilateraltimeout": kwargs.get("unilateraltimeout", 0),
    "destination": kwargs.get("destination", None)
}
return self.call("close", payload)
```

File: lightningd/peer_control.c [hunk 18]
```diff
@@ -1230,76 +1287,138 @@ static struct command_result *json_close(struct command *cmd,
...
+				/* Old default timeout */
+				if (!secondtok)
+					*timeout = 30;
+				else {
+					if (!json_to_number(buffer, secondtok, timeout))
+						return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+								    "close: Expected timeout to be a number. "
+								    "This argument ordering is deprecated!");
+				}
+			} else {
```
Comment (darosior) [inline lightningd/peer_control.c]: I think  you could merge the two `if ... else ...` ?

File: CHANGELOG.md [hunk 19]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   and has `connected` and `state` fields for channels, like `listpeers`.
 - JSON API: `fundchannel_start` now includes field `scriptpubkey`
 - JSON API: `txprepare` and `withdraw` now accept an optional parameter `utxos`, a list of utxos to include in the prepared transaction
+- JSON_API: `close` now accept an optional parameter `destination`, to which the to-local output will be sent.
```
Comment (darosior) [inline CHANGELOG.md]: nit: `s/accept/accepts/`

File: contrib/pylightning/lightning/lightning.py (hunk 17 above)
Comment (darosior) [inline contrib/pylightning/lightning/lightning.py]: Why do you use an embedded function ? I think you could use (did not test though)
```python
payload = {
    "id": peer_id,
    "unilateraltimeout": kwargs.get("unilateraltimeout", 0),
    "destination": kwargs.get("destination", None)
}
return self.call("close", payload)
```

File: channeld/channel_wire.csv (hunk 15 above)
Comment (trueptolemy) [reply channeld/channel_wire.csv]: Yes. For u8 array, `towire` will get this parameter by `tal_count`, and `fromwire` use it to initial a u8 array by `tal_array`. The bottom details are warpped in `wire/towire.c` and `wire/fromwire.c`.

File: lightningd/peer_control.c (hunk 18 above)
Comment (trueptolemy) [reply lightningd/peer_control.c]: Yeah, thank you! I merge these two code blocks into the one of `if(firsttok)` case.

File: contrib/pylightning/lightning/lightning.py (hunk 16 above)
Comment (trueptolemy) [reply contrib/pylightning/lightning/lightning.py]: Good catch!
Here should also adds:
```
            if len(args) == 2:
                if args[0] == None and isinstance(args[1], int):
                    return self._deprecated_close(peer_id, *args, **kwargs)
```

File: contrib/pylightning/lightning/lightning.py (hunk 17 above)
Comment (trueptolemy) [reply contrib/pylightning/lightning/lightning.py]: This embedded function is the "real" new-style `close`, comparing with `_deprecated_close`. Do you think this explanation is ok?

Comment (darosior) [reply contrib/pylightning/lightning/lightning.py]: Yeah it's just a code style choice not blocking anything :-)

Comment (trueptolemy) [reply contrib/pylightning/lightning/lightning.py]: Thank you very much! :-)

**Comment by niftynei** [comment]
ACK https://github.com/ElementsProject/lightning/pull/3116/commits/390b6b7c7873a386a27e866bd6ededa0cfc9985d

Added a small change to update the short help text since it was sorely out of date. tested cli interface -- looks good.

**Comment by niftynei** [comment]
ACK https://github.com/ElementsProject/lightning/pull/3116/commits/09aa603b3ba71f5fa193f15ea0229e8b610dedc2

**Pull request merged**